plutus-core/plutus-ir/test/TypecheckSpec.hs

```diff
@@ -0,0 +1,141 @@
+tyCheckPure ::
+  -- | The pure simplification pass.
+  (Term TyName Name DefaultUni DefaultFun () -> Term TyName Name DefaultUni DefaultFun ())
+  -> Term TyName Name DefaultUni DefaultFun ()
+  -> Either (Error DefaultUni DefaultFun ()) ()
+tyCheckPure pass tm = do
+  config <- getDefTypeCheckConfig ()
+  inferredType <- runQuoteT $ inferType config tm
+  runQuoteT $ checkType config () (pass tm) inferredType
+
+prop_typecheck_pure ::
+  (Term TyName Name DefaultUni DefaultFun ()
+  -> Term TyName Name DefaultUni DefaultFun ())
+  -> Property
+prop_typecheck_pure pass =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) -> do
+  convertToEitherString $ tyCheckPure pass tm
+
+-- | Typechecking a term after a `PlutusIR.Compiler.simplifyTerm` pass.
+prop_typecheck :: Property
+prop_typecheck =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
+    convertToEitherString $ do
+    config <- getDefTypeCheckConfig ()
+    plcConfig <- PLC.getDefTypeCheckConfig ()
+    inferredType <- runQuoteT $ inferType config tm
+    case runReaderT (runQuoteT (simplifyTerm (original tm))) (toDefaultCompilationCtx plcConfig) of
+      Left err -> Left $ void err
+      Right simplified ->
+        runQuoteT $ checkType config () (void simplified) inferredType
+
+-- | Typechecking a term after a `PlutusIR.Transform.StrictifyBindings` pass.
+prop_typecheck_strictifyBindings :: BuiltinSemanticsVariant DefaultFun -> Property
+prop_typecheck_strictifyBindings biVariant =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
+    convertToEitherString $ do
+    config <- getDefTypeCheckConfig ()
+    inferredType <- runQuoteT $ inferType config tm
+    let strictified = strictifyBindings (BuiltinsInfo biVariant) (original tm)
+    runQuoteT $ checkType config () (void strictified) inferredType
+
+-- | Typechecking a term after a `PlutusIR.Transform.EvaluateBuiltins` pass.
+prop_typecheck_evaluateBuiltins :: BuiltinSemanticsVariant DefaultFun -> Property
+prop_typecheck_evaluateBuiltins biVariant =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
+    convertToEitherString $ do
+    config <- getDefTypeCheckConfig ()
+    inferredType <- runQuoteT $ inferType config tm
+    let evaluated = evaluateBuiltins True (BuiltinsInfo biVariant) def (original tm)
+    runQuoteT $ checkType config () (void evaluated) inferredType
+
+-- | Typechecking a term after a `PlutusIR.Transform.Inline` pass.
+prop_typecheck_inline :: BuiltinSemanticsVariant DefaultFun -> Property
+prop_typecheck_inline biVariant =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
```

Comment:
Does the generated term have globally unique names? If not, we need to rename before we start otherwise the inliner will not work. (Yes, we're totally inconsistent in whether or not we rename inside the pass or not).

Reply:
Good point. I'm just gonna `rename` to be sure. Doesn't seem to make the tests pass though.